pkg/controller/elasticsearch/client/error.go

```diff
@@ -0,0 +1,92 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License 2.0;
+// you may not use this file except in compliance with the Elastic License 2.0.
+
+package client
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+)
+
+// APIError is a non 2xx response from the Elasticsearch API
+type APIError struct {
+	Status        string
+	StatusCode    int
+	ErrorResponse ErrorResponse
+}
+
+// newAPIError converts an HTTP response into an API error, attempting to parse the body to include the details about the error.
+func newAPIError(response *http.Response) error {
+	defer response.Body.Close()
+	apiError := &APIError{
+		Status:     response.Status,
+		StatusCode: response.StatusCode,
+	}
+	// We may need to read the body multiple times, read the full body and store it as an array of bytes.
+	body, err := ioutil.ReadAll(response.Body)
```

Comment:
nit: I would put the `defer response.Body.Close() ` statement here (right after we decide to use it, don't forget to close it).

Reply:
Maybe I'm missing something, the response body is always read by `body, err := ioutil.ReadAll(response.Body)`, there is no condition. This is why the `defer response.Body.Close()` is right before.

Reply:
I think by convention people tend to put the `defer` statement immediately after the line that requires it ([effective go example](https://golang.org/doc/effective_go#defer), [uber go style example](https://github.com/uber-go/guide/blob/master/style.md#defer-to-clean-up)), but I really don't mind much.

Reply:
I think the situation is a bit different here since the stream is already opened. The purpose is to be sure that it is effectively closed when we exit the function and no one else will read from it later.